### ISSUE TITLE
ABLoss Implentation

### ISSUE THREAD

**Issue by xd17** [bhheo/AB_distillation#6]
![image](https://user-images.githubusercontent.com/32261903/109413770-22e9fc00-79ea-11eb-9902-1e76b3cd2f88.png)
` loss_AT4 = ((self.Connect4(res4_s) > 0) ^ (res4_t > 0)).sum().float() / res4_t.nelement()`
`            loss = self.criterion_active_L2(self.Connect4(res4_s), res4_t.detach(), margin) / self.batch_size
`
 I read your paper, this is a good work.   but I am a little  confused about the implementation, I could not find the corresponding implementation of Formula 7 in your paper. Could your give me a explanation, waiting your reply. thanks

**Comment by bhheo** [comment]
Hi

`loss_AT` is only used for loss visualization

You can find the loss function for back-propagation here
https://github.com/bhheo/AB_distillation/blob/a7701d3506e13da87bda75aba1cee6397408144f/cifar10_AB_distillation.py#L111-L113

and `criterion_active_L2` is here https://github.com/bhheo/AB_distillation/blob/a7701d3506e13da87bda75aba1cee6397408144f/cifar10_AB_distillation.py#L16-L20

In detail,
`r(S(I))` is `d_net.Connectx(d_net.resx)`
`\rho(T(I))` is `(target > 0)` and `1 - \rho(T(I))` is `(target <= 0)`
`(source + margin)**2 * ((source > -margin)` conduct role of `\sigma(\mu 1 + r(S(I)))` and L2 norm

I hope it helps you to understand the implementation
